core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -1087,12 +1087,14 @@ class Partition(val topicPartition: TopicPartition,
     // avoid unnecessary collection generation
     val leaderLogEndOffset = leaderLog.logEndOffsetMetadata
     var newHighWatermark = leaderLogEndOffset
-    remoteReplicasMap.values.foreach { replica =>
+    remoteReplicasMap.foreachEntry { (replicaId, replica) =>
```

Comment:
can we use `remoteReplicasMap.values` here and use the replica.brokerId similar to the maximalIsr.contains call?

Reply:
Should we have a test in `PartitionTest` to assert that the HWM is incremented when there is a replica that is fenced but caught up?

Reply:
Done and Done. I confirmed that the test I added fails for the "fenced" and "shutdown" variant against trunk.